devito/passes/iet/misc.py

```diff
@@ -90,23 +93,96 @@ def relax_incr_dimensions(iet, **kwargs):
         outer, inner = split(iterations, lambda i: not i.dim.parent.is_Block)
 
         # Get root's `symbolic_max` out of each outer Dimension
+        roots_min = {i.dim.root: i.symbolic_min for i in outer}
         roots_max = {i.dim.root: i.symbolic_max for i in outer}
 
         # Process inner iterations and adjust their bounds
-        for n, i in enumerate(inner):
-            # The Iteration's maximum is the MIN of (a) the `symbolic_max` of current
-            # Iteration e.g. `x0_blk0 + x0_blk0_size - 1` and (b) the `symbolic_max`
-            # of the current Iteration's root Dimension e.g. `x_M`. The generated
-            # maximum will be `MIN(x0_blk0 + x0_blk0_size - 1, x_M)
+        for i in inner:
+            # Usually the Iteration's maximum is the MIN of (a) the `symbolic_max` of
+            # current Iteration e.g. `x0_blk0 + x0_blk0_size - 1` and (b) the
+            # `symbolic_max` of the current Iteration's root Dimension e.g. `x_M`. The
+            # generated maximum will be `MIN(x0_blk0 + x0_blk0_size - 1, x_M)
 
             # In some corner cases an offset may be added (e.g. after CIRE passes)
             # E.g. assume `i.symbolic_max = x0_blk0 + x0_blk0_size + 1` and
             # `i.dim.symbolic_max = x0_blk0 + x0_blk0_size - 1` then the generated
             # maximum will be `MIN(x0_blk0 + x0_blk0_size + 1, x_M + 2)`
 
+            # Step 1: Manage interval offsets
+            # Here we create two dictionaries to keep track of offsets incurring in
+            # 'i' and 'i.root' iteration
+            root_min = roots_min[i.dim.root] + i.symbolic_min - i.dim.symbolic_min
+            min_map = {
+                i.dim.root.symbolic_min: root_min,
+                i.dim.symbolic_min: i.symbolic_min
+            }
+
             root_max = roots_max[i.dim.root] + i.symbolic_max - i.dim.symbolic_max
-            iter_max = evalrel(min, [i.symbolic_max, root_max])
-            mapper[i] = i._rebuild(limits=(i.symbolic_min, iter_max, i.step))
+            max_map = {
+                i.dim.root.symbolic_max: root_max,
+                i.dim.symbolic_max: i.symbolic_max
+            }
+
+            # Step 2: Generate assumptions deriving from hierarchical blocking. Assuming
```

Comment:
Note that more than "assuming", it is "by construction" (ie, user should get an exception if he/she tries to set the mid BlockDimension to anything that is not a perfect division of the parent block size... something like this. I hope this is enforced somewhere, perhaps at args processing time?)

Reply:
This is kinda not-related to this part of code? I mean it is something irrelevant of this PR. (right?)
So maybe it just deserves another small PR?